### DISCUSSION TITLE
A compact internal representation for record accessors

### DISCUSSION THREAD

**Discussion by gbaz** [ghc-proposals/ghc-proposals#470] [Ideas]
This is an issue rather than a PR. I don't know enough about the guts of GHC and the ease or difficulty of various approaches to suggest a full solution, and I'm hoping for ghc experts to weigh in. I've been pondering the problems described in well-typed's work on large records: https://well-typed.com/blog/2021/08/large-records/

If we only consider the simplest example -- quadratic blowup in code size for accessors, this comes from the fact that accessors are generated from all fields, and also need to enumerate all fields. Given how core currently looks, this seems pretty inevitable and natural. My question is the following: is there a way to change core to have some "magic" notion of a compact accessor that serves as sugar for this? It seems to me also that accessors should invariably be inlined. So could we just have a magic accessor token in core that is always expanded _at use_ -- i.e. a compact representation that "desugars" to the full thing inlined thing when invoked, but is otherwise efficient to generate and keep around?

Personally, I think solving issues like applicative, etc. is less of a key case, since most fields in the autogenerated big records I am used to are not parameterized. I haven't thought through the other generated typeclasses enough to consider which ones may or may not present similar quadratic issues -- experience makes me wary of Read/Show/Eq/Ord, but logically I don't think a number of them would be subject to these problems.

**Comment by nomeata** [comment]
This sounds more like a discussion than an issue, so I took the liberty to convert.

**Comment by Ericson2314** [comment]
Does `-XNoFieldAccessors` bypass this issue (making it just the `O(n)` of patterns the user actually wrote)? While I think it's good to fix this in any event, `-XNoFieldAccessors` helping would take a little bit of the pressure off for me.

**Reply by adamgundry** [reply]
Unfortunately not in current GHC versions, because even with `NoFieldSelectors`, the functions are still generated in the background (for use solving `HasField` constraints). I'm planning (one day) to propose knobs to tune this (i.e. control whether the functions get generated at record datatype definition sites or at selector/`HasField` use sites).

**Comment by simonpj** [comment]
The question arises: is it better 
1.  to generate 100 functions at the data type defintion, that are used many times, or
2. generate no function at the data type defintion, but instead inline a copy of the function at every use site

Well, if there are no use sites, then (2) is better.  If there are many use sites, maybe (1) is better.   But it is hard to know in advance, because it depends on the not-yet-written uses.

Mind you, the selector is often inlined, so (1) actually behaves much like (2).   This is the "knob" that Adam refers to.

**Comment by gbaz** [comment]
Pardon the levity, but Simon's phrasing reminds me of the old question: "Would you rather fight 1 horse-sized duck or 100 duck-sized horses?"

I'm particularly interested in any metrics from @edsko or others that might indicate if resolving this selector quadratic blowup alone would be of significant use in the cases of large records they examined. I'll try to slice off some time in the next few days and examine the core in our work repos to see where the blowup is most significant as well...

I will say that in my experience, not having any particular selector at hand can be a real pain, but most generated selectors are not used, most of the time, but also having named record fields is pretty essential to being able to derive all the other instances we want for a datatype, and otherwise use it sanely.

**Comment by Ericson2314** [comment]
> Mind you, the selector is often inlined, so (1) actually behaves much like (2). This is the "knob" that Adam refers to.

Yeah I feel like this might be a good enough default that we don't need a job. Long term, if we also have the stuff we need for https://gitlab.haskell.org/ghc/ghc/-/issues/14917, we might be able to dedup any residual non-inlined selectors.

**Comment by edsko** [comment]
@simonpj writes:

> The question arises: is it better
> * to generate 100 functions at the data type defintion, that are used many times, or
> * generate no function at the data type defintion, but instead inline a copy of the function at every use site

but I feel this is a false dichotomy. The problem is not that we generate 100 functions at the data type definition. The problem is that each of those functions is itself linear. If we had a better representation for record access and record update in core, this problem would disappear, Personally, I think this is the right solution, and of all quadratic problems that I've investigated, the easiest one to fix. The "only" problem of course is that it would necessitate a change to core.

**Comment by simonpj** [comment]
> The "only" problem of course is that it would necessitate a change to core.

Ah yes -- but what change did you have in mind.   Woudl that change fix other quadratic problems?

**Comment by edsko** [comment]
We'd need the core equivalent of `\Foo{x} -> x` and `\foo -> foo {x == ..}`, without mentioning all the other variables. As regards other quadratic problems: the same could also be used for dictionaries, which is important at least in the case of classes with lots of superclass constraints. I'm not aware of any other problems this would fix off the top of my head. In a sense, it's the least interesting of the quadratic problems, and possibly even the least critical, but probably also the lowest hanging fruit.

**Reply by treeowl** [reply]
I'm pretty sure that first one should be more like `\Foo{x} -> (# x #)`.

**Comment by adamgundry** [comment]
I'm wondering if it would make sense to extend Core's `case` to give a compact representation to particular lists of alternatives, in particular "select the field `x`" and "update the field `x` to value `v`".

That is, instead of `\Foo{x} -> x` desugaring to `\v -> case v of { Foo x ... -> x }` it would desugar to `\v -> case v of select "x"`.

Similarly, `\foo -> foo { x = v }` currently desugars to `\foo -> case foo of { Foo _ y2 ... yn -> Foo v y2 ... yn }` but it could instead be `\foo -> case foo of update "x" v`.

In code, we'd change the existing `Expr` constructor
```hs
data Expr b = ...
 | Case  (Expr b) b Type [Alt b]
```
to be something like
```hs
 | Case  (Expr b) b Type (Alts b)

data Alts b = Normal [Alt b] | Select FieldLabel | Update FieldLabel (Expr b)
```

I'm not sure how easy it would be to preserve this representation during optimization, though.

**Comment by simonpj** [comment]
Doing something like this would disable optimisation.  Consider
```
case x of 
  K a _b _c -> .....(case x of K _p q _r -> rhs)....
````
and suppose that each of those cases picks only field, so the underscored variables are not mentioned. 

GHC will add the info that `x = K a _b _c` to its environment when processing the outer case alternative.  Then it'll simply the inner `case x`, since it knows what x is.  So we get
```
case x of
  K a _b _c -> ...(rhs[ a/_p, _b/q, c/_r])...
```
substituting `a` for `_p` etc.

One we are done, `_b` is alive, whereas before it was dead.  If the outer case didn't even bind `_b` (because it has become a magic record selector) we could not do this.

**Reply by adamgundry** [reply]
Sorry, I don't quite follow the problem here. With the representation I suggested in https://github.com/ghc-proposals/ghc-proposals/discussions/470#discussioncomment-1763384 the outer case cannot be represented using the new form (because its alternative isn't a select or update). And I don't think there's a problem representing the inner case using the new form (if `rhs` is just `q`, for example) as it can still be simplified using `x = K a _b _c`. But perhaps you had a different representation in mind?

My assumption would be that we could pick a representation that encodes all the same information as the fully expanded `case`, and hence can be expanded during optimization if necessary. Thus optimization might lead to loss of the compact representation, but the compact representation shouldn't block runtime optimizations.

**Comment by simonpj** [comment]
Oh I see.. your alternative embodies the RHS too.  But then immediately we had
```
case (case v of SELECT x) of alts
```
we'll have to fluff up the inner case to allow the case-of-case transformation to work:
```
case v of
  K _ _ _ _ x _ _ _ -> case x of alts
```
So the selector rapidly gets blown up to be what it is now.   To me it looks like a lot of new code in GHC, for little or no benefit.

**Comment by Ericson2314** [comment]
I could imagine we have a new shorthand constructor for these sorts of variants, and then a clever pattern synonym that hides the details?

It still makes me nervous to so "denormalize" the Core AST, but it ought to be something we can do safely.

**Comment by jcarr-awake** [comment]
Could this be implemented by instead changing each of the constructors in a case alternative to be sparse lists of arguments? (conceptually a `[(Int, b)]`)
This would add some complexity to the compiler, but would largely keep Core looking "the same".

I.e. all of our single-constructor case constructs would look like:
```
case v of
   K {1=a, 5=b ...} -> body
```

In that case, we could handle Simon's optimization, just by:
1. creating new names on the outer construct for each field accessed in the inner case block
2. applying the optimization with the new outer case block

```
case x of 
  K {1=a, ...} -> .....(case x of K {2=q, ...} -> rhs)....
```
--->
```
case x of 
  K {1=a, 2=b, ...} -> .....(rhs[b/q])....
```

I believe updates could be handled the same way? As a sparse, ordered set of fields to update. Or should we only update one at a time?
